core/src/main/scala/kafka/server/ConfigHelper.scala

```diff
@@ -47,11 +47,11 @@ class ConfigHelper(metadataCache: MetadataCache, config: KafkaConfig, configRepo
 
       def createResponseConfig(configs: Map[String, Any],
                                createConfigEntry: (String, Any) => DescribeConfigsResponseData.DescribeConfigsResourceResult): DescribeConfigsResponseData.DescribeConfigsResult = {
-        val filteredConfigPairs = if (resource.configurationKeys == null)
+        val filteredConfigPairs = if (resource.configurationKeys == null || resource.configurationKeys.isEmpty)
```

Comment:
This change doesn't seem correct.  null does not mean the same thing as empty.

This is how `ConfigurationKeys` is described in the schema:
```
      { "name": "ConfigurationKeys", "type": "[]string", "versions": "0+", "nullableVersions": "0+",
        "about": "The configuration keys to list, or null to list all configuration keys." }
```
Notice it does not say "null or empty to list all configuration keys."  Just null.

Also, I looked at the code from the 2.7 release and it does not treat empty the same as null.

Reply:
I agree with this. However it seems treating empty as null is de facto behavior at the moment (before the fix). To be more precise, here
https://github.com/apache/kafka/blob/e2a0d0c90e1916d77223a420e3595e8aba643001/core/src/main/scala/kafka/server/ConfigHelper.scala#L53-L55
`resource.configurationKeys` being empty means `true` for each element of `configs`, so effectively no filtering. Which, in turn, is equivalent to `resource.configurationKeys == null`.

For example, `testDescribeConfigsWithDocumentation` and `testDescribeConfigsWithEmptyConfigurationKeys` in `ZkAdminManagerTest` break after the fix if treating empty as null is not kept.

I guess it originates from that the default constructor of `DescribeConfigsResource` makes `configurationKeys` an empty array.

We can either 1) keep the current behavior (the current approach); or 2) stick to the spec (also potentially adding `"default": "null"` to the field description). I'd prefer 2. What do you think?

Reply:
As you said earlier, the code in 2.7 is broken.  So we should't be going based off of that.

However, it occurs to me that the behavior is the same whether we special case the empty string or not. Since every string contains the empty string, specifying the empty string will have the same effect as specifying null. This is a good example of a case where a field shouldn't have been nullable.

Anyway, can you take out the special case? It seems likely to create confusion.